Enable xray support for Mac

quininer opened the pull request:
https://github.com/rust-lang/rust/issues/102921 

Upstream has supported Mac for a while, let's enable it.

I've tested it on M4 and it generates nop sled correctly.

* https://maskray.me/blog/2023-06-18-port-llvm-xray-to-apple-systems
* https://github.com/llvm/llvm-project/blob/llvmorg-20.1.4/clang/lib/Driver/XRayArgs.cpp#L31

try-job: x86_64-apple-1
try-job: aarch64-apple

quininer committed:
Enable xray support for Mac

* https://maskray.me/blog/2023-06-18-port-llvm-xray-to-apple-systems
* https://github.com/llvm/llvm-project/blob/llvmorg-20.1.4/clang/lib/Driver/XRayArgs.cpp#L31

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14904068753/job/41862404963) [(plain)](https://github.com/rust-lang/rust/commit/e8d351041bbfa80e74eecf2ebd85236f3a3b8005/checks/41862404963/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 19.9s done
#18 DONE 22.7s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---

error: ui test did not emit an error
note: by default, ui tests are expected not to compile
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/instrument-xray/target-not-supported.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/instrument-xray/target-not-supported" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Z" "instrument-xray" "--target" "x86_64-apple-darwin"
stdout: none
stderr: none



```

</details>

jieyouxu commented on tests/ui/instrument-xray/target-not-supported.rs:
```diff
@@ -1,7 +1,7 @@
 //
 //@ needs-llvm-components: x86
-//@ compile-flags: -Z instrument-xray --target x86_64-apple-darwin
+//@ compile-flags: -Z instrument-xray --target x86_64-pc-windows-msvc
```
Question: do we have any smoke test for xray on `x86_64-apple-darwin` and `aarch64-apple-darwin`? As in:

- A smoke test that have some ability to observe xray being enabled or not, for `aarch64-apple-darwin` and `x86_64-apple-darwin` targets.
- A test check that the flag `-Z instrument-xray` is still unstable, but is no longer rejected on `x86_64-apple-darwin` and `aarch64-apple-darwin` respectively.

quininer replied:
I think both of these tests don't exist at the moment, I can add tests for this but there are two problems

1. It seems that only llir is tested in `tests/codegen` and no asm, which cannot be used to check whether xray is enabled, because xray affects llvm output rather than input. i need some guidance how to write tests for asm output.
2. Add a test specifically for `instrument-xray` + `x86_64-apple-darwin`/`aarch64-apple-darwin` combination seems a bit strange to me, since this is not a feature we implement, but just an upstream feature. do we need to add such a test for each supported target?

jieyouxu replied:
1. There's also `tests/assembly` which does exercise assembly.
2. What I meant was just cli smoke tests to check that we properly gate the stability given the target platform. Whether the impl works or not is a separate question.

jieyouxu commented:
Since this affects the Tier 1 `{x86_64,aarch64}-apple-darwin` targets (albeit `-Z instrument-xray` is still unstable), asking Apple experts for second opinion as I neither have access to apple envs nor am I aware of how well LLVM xray currently works on those targets.

@rustbot ping apple

jieyouxu commented:
@rustbot author (test nits)

quininer committed:
Add compiler tests for xray

quininer commented:
I added tests for assembly output, which should be seen as supplementing the previous missing tests for xray.

I also added flag stable test, but I still don't think it's necessary. I don't see any other flags have this test, I can only get this from https://github.com/rust-lang/rust/blob/1.87.0/tests/ui/bootstrap/rustc_bootstrap.rs#L36 .
and this test has nothing to do with xray, but is a test of the order of `-Z` flag and target checks.

@rustbot ready

jieyouxu reviewed (approved):
Thanks, just a nit on the platform support test

madsmtm reviewed (commented):
As someone with very little experience with XRay, I tried to test this end-to-end:

```console
$ ./x build --set=build.sanitizers=true
$ echo "fn main() {}" > main.rs
$ rustc +stage1 -Zinstrument-xray=always main.rs
$ objdump -h -j xray_instr_map main # Works
$ XRAY_OPTIONS="patch_premain=true xray_mode=xray-basic verbosity=1" ./main # No-op?
```

But it doesn't seem to do anything? Probably because I'm not linking the runtime, but it's not clear to me how I'd go about doing that? Does `bootstrap` need to build it?

quininer commented:
@madsmtm As with linux, you need to link to `compiler-rt/xray`. like https://crates.io/crates/clang-rt-xray

Anyway I don't really care about that, because I'll be using my own rt, like [uftrace](https://github.com/namhyung/uftrace). So just rustc generate the correct nop sled and `xray_instr_map` section will good.

jieyouxu commented:
> As with linux, you need to link to `compiler-rt/xray`. like [crates.io/crates/clang-rt-xray](https://crates.io/crates/clang-rt-xray)
> 
> Anyway I don't really care about that, because I'll be using my own rt, like [uftrace](https://github.com/namhyung/uftrace?rgh-link-date=2025-05-19T16%3A35%3A06Z). So just rustc generate the correct nop sled and `xray_instr_map` section will good.

Could you add that as a remark on the unstable docs for this flag?

quininer commented on tests/ui/instrument-xray/target-supported-but-unstable.rs:
```diff
@@ -0,0 +1,9 @@
+//@ rustc-env:RUSTC_BOOTSTRAP=-1
+//@ compile-flags: -Z instrument-xray --target x86_64-unknown-linux-gnu
+
+#![feature(no_core)]
+#![no_core]
+#![no_main]
```
Ok, I just copy this from `tests/ui/instrument-xray/target-not-supported.rs`.

quininer replied:
This does not seem to work, which would cause the test to fail on linux. (because darwin std cannot be built).

If we are just test flag, using `no_main` seems reasonable to me.

jieyouxu replied:
Ah right, yeah that's fine.

quininer commented:
> Could you add that as a remark on the unstable docs for this flag?

@jieyouxu I think it is already mentioned in  unstable book. https://github.com/rust-lang/rust/blob/1.87.0/src/doc/unstable-book/src/compiler-flags/instrument-xray.md?plain=1#L35

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15127580742/job/42522382758) [(plain)](https://github.com/rust-lang/rust/commit/fbfb21013542b8a9a435624f1b3e0b1af27cd58c/checks/42522382758/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.915 Building wheels for collected packages: reuse
#12 2.916   Building wheel for reuse (pyproject.toml): started
#12 3.131   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.132   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#12 3.132   Stored in directory: /tmp/pip-ephem-wheel-cache-4bj7c4l7/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.134 Successfully built reuse
#12 3.135 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.548 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.548 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.094 Collecting virtualenv
#12 4.154   Downloading virtualenv-20.31.2-py3-none-any.whl (6.1 MB)
#12 4.389      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.1/6.1 MB 26.0 MB/s eta 0:00:00
#12 4.450 Collecting platformdirs<5,>=3.9.1
#12 4.489   Downloading platformdirs-4.3.8-py3-none-any.whl (18 kB)
#12 4.532 Collecting filelock<4,>=3.12.2
#12 4.544   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.567 Collecting distlib<1,>=0.3.7
#12 4.581   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.593      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 44.7 MB/s eta 0:00:00
#12 4.675 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.868 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.8 virtualenv-20.31.2
#12 4.869 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 5.0s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      151488 kB
DirectMap2M:     8237056 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-05-12/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
[TIMING] core::build_steps::tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 6022 files
tidy check
/checkout/tests/ui/instrument-xray/platform-support.rs: revision unsupported should specify `needs-llvm-components:` as it has `--target` set
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.1.1)
linting python files
All checks passed!
checking python file formatting
26 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:57
  local time: Tue May 20 02:46:50 UTC 2025
  network time: Tue, 20 May 2025 02:46:50 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

quininer commented:
@rustbot ready

jieyouxu reviewed (approved):
Thanks

jieyouxu commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 4a619fb960a432c4773395d5f32b359f83b28f54 has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu 4a619fb960a432c4773395d5f32b359f83b28f54 -->
<!-- homu: {"type":"Approved","sha":"4a619fb960a432c4773395d5f32b359f83b28f54","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r- https://github.com/rust-lang/rust/pull/141473#issue-3087606050

jieyouxu commented:
@bors r+ rollup=iffy

bors commented:
:pushpin: Commit ac00d79d6ffb4b8593f93b89fbcff248cc7339fa has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu ac00d79d6ffb4b8593f93b89fbcff248cc7339fa -->
<!-- homu: {"type":"Approved","sha":"ac00d79d6ffb4b8593f93b89fbcff248cc7339fa","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit ac00d79d6ffb4b8593f93b89fbcff248cc7339fa with merge d1037a7492c454509b1338be9d1144d3c8ecdad4...
<!-- homu: {"type":"BuildStarted","head_sha":"ac00d79d6ffb4b8593f93b89fbcff248cc7339fa","merge_sha":"d1037a7492c454509b1338be9d1144d3c8ecdad4"} -->

rust-log-analyzer commented:
The job **`x86_64-apple-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15227384059/job/42831098091) [(plain)](https://github.com/rust-lang/rust/commit/d1037a7492c454509b1338be9d1144d3c8ecdad4/checks/42831098091/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [assembly] tests/assembly/x86_64-xray.rs stdout ----

error: compilation failed!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/Users/runner/work/rust/rust/build/x86_64-apple-darwin/stage2/bin/rustc" "/Users/runner/work/rust/rust/tests/assembly/x86_64-xray.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/Users/runner/.cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/Users/runner/work/rust/rust/vendor" "--sysroot" "/Users/runner/work/rust/rust/build/x86_64-apple-darwin/stage2" "--check-cfg" "cfg(test,FALSE)" "-O" "-Cdebug-assertions=no" "--emit" "asm" "-C" "prefer-dynamic" "-o" "/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/assembly/x86_64-xray/x86_64-xray.s" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "--crate-type" "lib" "-Zinstrument-xray=always" "-C" "llvm-args=-x86-asm-syntax=intel" "--target" "x86_64-unknown-linux-gnu"
stdout: none
--- stderr -------------------------------
error[E0463]: can't find crate for `std`
  |
  = note: the `x86_64-unknown-linux-gnu` target may not be installed
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/rust/actions/runs/15227384059/job/42832248341)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust/actions/runs/15227384059/job/42832248341","builder_name":"checks-actions"} -->

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit 5dfd723e54e2e786e71174fdedc9f7c04facb157 with merge ac496294fce7b3f999efb79bdc5adcb319d6b2f8...
<!-- homu: {"type":"TryBuildStarted","head_sha":"5dfd723e54e2e786e71174fdedc9f7c04facb157","merge_sha":"ac496294fce7b3f999efb79bdc5adcb319d6b2f8"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15228381537/job/42833230835) [(plain)](https://github.com/rust-lang/rust/commit/5dfd723e54e2e786e71174fdedc9f7c04facb157/checks/42833230835/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 21.1s done
#18 DONE 27.8s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
---- [assembly] tests/assembly/x86_64-xray.rs stdout ----

error: compilation failed!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/assembly/x86_64-xray.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--check-cfg" "cfg(test,FALSE)" "-O" "-Cdebug-assertions=no" "--emit" "asm" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/x86_64-xray/x86_64-xray.s" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Zinstrument-xray=always" "-C" "llvm-args=-x86-asm-syntax=intel" "--target" "x86_64-unknown-linux-gnu" "-Cpanic=abort" "-Cforce-unwind-tables=yes" "--extern" "minicore=/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/x86_64-xray/libminicore.rlib"
stdout: none
--- stderr -------------------------------
error[E0433]: failed to resolve: use of unresolved module or unlinked crate `std`
##[error]  --> /checkout/tests/assembly/x86_64-xray.rs:15:5
   |
---
##[error]  --> /checkout/tests/assembly/x86_64-xray.rs:12:20
   |
12 |   pub fn xray_func() {
   |  ____________________^
13 | |     // CHECK: nop word ptr [rax + rax + 512]
14 | |
15 | |     std::hint::black_box(());
...  |
19 | | }
   | |_^
```

</details>

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15228574352/job/42833622754) [(plain)](https://github.com/rust-lang/rust/commit/ac496294fce7b3f999efb79bdc5adcb319d6b2f8/checks/42833622754/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling ureq v3.0.8
   Compiling citool v0.1.0 (/home/runner/work/rust/rust/src/ci/citool)
    Finished `dev` profile [unoptimized] target(s) in 6.91s
     Running `target/debug/citool calculate-job-matrix`
Run type: TryJob { job_patterns: Some(["x86_64-apple-1", "aarch64-apple-1"]) }
Error: Failed to calculate job matrix

Caused by:
    Patterns `aarch64-apple-1` did not match any auto jobs
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

jieyouxu commented on tests/assembly/x86_64-xray.rs:
```diff
@@ -0,0 +1,19 @@
+pub fn xray_func() {
+    // CHECK: nop word ptr [rax + rax + 512]
+
+    std::hint::black_box(());
```
Ah of course, you can't use this if `no_core`...

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit 5171fc6136379d9e39cb39de6ecf2bededc20f1d with merge d71bebabb9e671415292f04e1537b77e2916557d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"5171fc6136379d9e39cb39de6ecf2bededc20f1d","merge_sha":"d71bebabb9e671415292f04e1537b77e2916557d"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15228817795/job/42834169833) [(plain)](https://github.com/rust-lang/rust/commit/5171fc6136379d9e39cb39de6ecf2bededc20f1d/checks/42834169833/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 20.5s done
#18 DONE 27.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
---- [assembly] tests/assembly/x86_64-xray.rs stdout ----

error: compilation failed!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/assembly/x86_64-xray.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--check-cfg" "cfg(test,FALSE)" "-O" "-Cdebug-assertions=no" "--emit" "asm" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/x86_64-xray/x86_64-xray.s" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Zinstrument-xray=always" "-C" "llvm-args=-x86-asm-syntax=intel" "--target" "x86_64-unknown-linux-gnu" "-Cpanic=abort" "-Cforce-unwind-tables=yes" "--extern" "minicore=/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/x86_64-xray/libminicore.rlib"
stdout: none
--- stderr -------------------------------
error: requires `sized` lang_item
##[error]  --> /checkout/tests/assembly/x86_64-xray.rs:12:29
   |
12 | pub fn xray_func(n: u32) -> u32 {
   |                             ^^^

error: aborting due to 1 previous error
------------------------------------------

```

</details>

quininer commented:
Well, I don't have env to really test it right now, so let me look at it tomorrow. :'(

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/rust/actions/runs/15228875392/job/42834984606)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang/rust/actions/runs/15228875392/job/42834984606","builder_name":"checks-actions"} -->

quininer commented on tests/assembly/x86_64-xray.rs:
```diff
@@ -0,0 +1,19 @@
+#![feature(no_core)]
+#![no_core]
+#![crate_type = "lib"]
+
```
I still tried using `only-linux` flag in order to test in most normal scenarios.

bors commented:
:pushpin: Commit a5c42a5ada041d84b69ae9db1af5802f4b3183a7 has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu a5c42a5ada041d84b69ae9db1af5802f4b3183a7 -->
<!-- homu: {"type":"Approved","sha":"a5c42a5ada041d84b69ae9db1af5802f4b3183a7","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15236996933/job/42852183862) [(plain)](https://github.com/rust-lang/rust/commit/08c9e3c73c18e767aed3f8c495a1f8d03538bc53/checks/42852183862/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [assembly] tests/assembly/x86_64-xray.rs stdout ----

error: compilation failed!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/assembly/x86_64-xray.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--check-cfg" "cfg(test,FALSE)" "-O" "-Cdebug-assertions=no" "--emit" "asm" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/assembly/x86_64-xray/x86_64-xray.s" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Zinstrument-xray=always" "-C" "llvm-args=-x86-asm-syntax=intel" "--target" "x86_64-unknown-linux-gnu"
stdout: none
--- stderr -------------------------------
error[E0463]: can't find crate for `std`
  |
  = note: the `x86_64-unknown-linux-gnu` target may not be installed
```

</details>

madsmtm commented on tests/assembly/x86_64-xray.rs:
```diff
@@ -0,0 +1,17 @@
+//@ assembly-output: emit-asm
+//@ compile-flags: -Zinstrument-xray=always -C llvm-args=-x86-asm-syntax=intel --target x86_64-unknown-linux-gnu
+//@ needs-llvm-components: x86
+//@ only-x86_64-unknown-linux-gnu
```
But... Now we aren't actually testing this on macOS?

jieyouxu replied:
Hm yeah, shouldn't this assembly smoke test be for `x86_64-apple-darwin` and `aarch64-apple-darwin` 🤔

quininer replied:
I don't think it's that important, because x86_64 generation is not related to mac support.

madsmtm replied:
> I don't think it's that important, because x86_64 generation is not related to mac support.

Would still be nice to have a test that it actually works on macOS.

quininer replied:
Anyway I added one, as well as aarch64 (but I didn't really test it because don't have device on hand)

jieyouxu commented:
@bors try

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15238822991/job/42856298590) [(plain)](https://github.com/rust-lang/rust/commit/8b7e6f84398dcfcc774d746d5e944892bd9a2bf0/checks/42856298590/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 4.657   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.665      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 86.2 MB/s eta 0:00:00
#12 4.703 Collecting filelock<4,>=3.12.2
#12 4.723   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.761 Collecting platformdirs<5,>=3.9.1
#12 4.781   Downloading platformdirs-4.3.8-py3-none-any.whl (18 kB)
#12 4.867 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 5.066 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.8 virtualenv-20.31.2
#12 5.066 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 5.2s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      149440 kB
DirectMap2M:     7190528 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-05-12/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
[TIMING] core::build_steps::tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 6029 files
tidy check
/checkout/tests/assembly/x86_64-xray.rs: revision [unspecified] should not specify `needs-llvm-components:` as it doesn't need `--target`
/checkout/tests/assembly/x86_64-xray.rs: revision x86_64-darwin should specify `needs-llvm-components:` as it has `--target` set
/checkout/tests/assembly/x86_64-xray.rs: revision x86_64-linux should specify `needs-llvm-components:` as it has `--target` set
/checkout/tests/assembly/aarch64-xray.rs: revision [unspecified] should not specify `needs-llvm-components:` as it doesn't need `--target`
/checkout/tests/assembly/aarch64-xray.rs: revision aarch64-darwin should specify `needs-llvm-components:` as it has `--target` set
/checkout/tests/assembly/aarch64-xray.rs: revision aarch64-linux should specify `needs-llvm-components:` as it has `--target` set
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.1.1)
linting python files
All checks passed!
checking python file formatting
26 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:46
  local time: Sun May 25 14:32:49 UTC 2025
  network time: Sun, 25 May 2025 14:32:49 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

jieyouxu commented:
@bors try

jieyouxu commented:
Oh right, sometimes there are CFI directives

quininer commented:
Ah, aarch64 generate sled in a different style than x86_64. It's always a jmp to ret instruction, not a direct ret.

I haven't really tested it so it may still have error.

bors commented:
:hourglass: Trying commit 45ed022d636ec72dfbb22269196007b0e71a9c7d with merge 1ad0547a1ba5d3926045afcd95e6fa17278ce9ba...
<!-- homu: {"type":"TryBuildStarted","head_sha":"45ed022d636ec72dfbb22269196007b0e71a9c7d","merge_sha":"1ad0547a1ba5d3926045afcd95e6fa17278ce9ba"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/15240019590/job/42861041412)
Build commit: 1ad0547a1ba5d3926045afcd95e6fa17278ce9ba (`1ad0547a1ba5d3926045afcd95e6fa17278ce9ba`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/15240019590/job/42861041412"},"merge_sha":"1ad0547a1ba5d3926045afcd95e6fa17278ce9ba"} -->

bors commented:
:pushpin: Commit 45ed022d636ec72dfbb22269196007b0e71a9c7d has been approved by `wesleywiser,jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=wesleywiser,jieyouxu 45ed022d636ec72dfbb22269196007b0e71a9c7d -->
<!-- homu: {"type":"Approved","sha":"45ed022d636ec72dfbb22269196007b0e71a9c7d","approver":"wesleywiser,jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/15263583756/job/42930612209)
Approved by: wesleywiser,jieyouxu
Pushing d76fe154029e03aeb64af721beafdcef856d576a to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"wesleywiser,jieyouxu","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/15263583756/job/42930612209"},"merge_sha":"d76fe154029e03aeb64af721beafdcef856d576a"} -->

quininer commented:
Thanks everyone for help, especially @jieyouxu 🎉
